typing/typecore.ml

```diff
@@ -2924,26 +2927,17 @@ and type_pat_aux
       let alloc_mode =
         cross_left !!penv expected_ty alloc_mode.mode
       in
+      let sort =
+        match Ctype.type_sort !!penv ty ~why:Let_binding ~fixed:false with
+        | Error _ -> assert false
+        | Ok sort -> sort
+      in
```

Comment:
I think that the `Error` case here might not be reachable right now, but it's basically just an accident that every path to pattern checking ensures the expected type is representable first, and this is likely to become false in the future. So the `Error` case should issue a proper error, and probably we should add a new `~why`, say `Pattern_variable`, for this case.

(As an aside, if we were going to keep this in its current form we should replace `assert false` with `Misc.fatal_error`. The former is OK in places where the invariant that makes something unreachable is very local, but where it depends on information from other functions and files it's better to use `Misc.fatal_error`, so we get a more helpful error if something changes that violates the invariant)